### ISSUE TITLE
High CPU on Single Core Virtual Machines

### ISSUE THREAD

**Issue by drewfogg** [Icinga/icinga2#7081]
## Describe the bug

Seems to be a similar issue to #5639

On Virtual Machines (VMWare ESXi) which have the Windows Icinga2 Agent installed and that also have a single CPU Core, the Icinga2 process maxes out the CPU rendering the system un-usable - seems to be quite aggressive, starving other processes of time on the CPU. While the process is pinned at 100%, the service checks are not running.

The problem can be resolved and then reproduced by adding and then removing CPU Cores. 

I found that if I leave Icinga2 maxing the CPU it eventually sorts itself out (can take anything from 10 to 60 minutes).

I also noticed that lowering the priority of the process using TaskManager or ProcessExplorer would almost instantly fix the process. Based on this I guessed that another process that Icinga2 is waiting for is trying to do something but going very slowly because of the lack of resources. It turns out that raising the priority of the child process of 'Icinga2.exe', 'conhost.exe', also sorts the issue.

I'm guessing its chicken and egg issue. Icinga2 is hanging because conhost.exe has not completed its task causing high CPU, and conhost.exe cannot complete its task because of the high CPU.

## To Reproduce

1. Setup a Windows VM
2. Install Icinga2 Windows Agent using PowerShell Script (https://github.com/Icinga/icinga2-powershell-module)
3. Once setup and working, change to CPU Cores to 1
4. Restart VM and when Icinga2 service starts the CPU should get pinned at 100%
5. Issue can be instantly resolved by doing either of the following:
 - Raising the priority of the conhost.exe process (the child process of Icinga2.exe)
 - Lowering the priroity of the Icinga2 process

## Screenshots
Viewing the Icinga2 Process maxing the CPU and changing the priority of conhost.exe using ProcessExplorer
![image](https://user-images.githubusercontent.com/13171415/55509829-cabaeb80-5654-11e9-9a36-97fab13eac4d.png)

## Your Environment
Include as many relevant details about the environment you experienced the problem in

Agents and Master(s) are on different subnets but there is no natting in either direction. 

We exclusively use the Icinga2 PowerShell Script for the agent installation.

* Version used (`icinga2 --version`):
Icinga2 Server: r2.10.4-1
Reproduced issue using v2.10.4 and v2.10.2 of the Windows Agent
* Operating System and version:
Icinga2 Server: 18.04.2 LTS 4.15.0-46-generic
Reproduced issue on:
   Windows Server 2012 R2 x64
   Windows Server 2012 R2 Core  x64
   Windows Server 2008 R2 x64
Wasn't able to reproduce on:
   Windows Server 2016 x64
* Enabled features (`icinga2 feature list`):
checker debuglog mainlog notification
* Icinga Web 2 version and modules (System - About):
director 1.6.2
doc 2.6.2
graphite 1.1.0
monitoring 2.6.2
* Config validation (`icinga2 daemon -C`):
```
[2019-04-03 21:03:45 +0100] information/cli: Icinga application loader (version: v2.10.4)
[2019-04-03 21:03:45 +0100] information/cli: Loading configuration file(s).
[2019-04-03 21:03:45 +0100] information/ConfigItem: Committing config item(s).
[2019-04-03 21:03:45 +0100] information/ApiListener: My API identity: *omitted*
[2019-04-03 21:03:46 +0100] warning/ApplyRule: Apply rule 'Disk space: /srv' (in C:\ProgramData\icinga2\var\lib\icinga2\
api\zones/director-global/director/servicesets.conf: 3:1-3:32) for type 'Service' does not match anywhere!
[2019-04-03 21:03:46 +0100] warning/ApplyRule: Apply rule 'Icinga 2: IDO status' (in C:\ProgramData\icinga2\var\lib\icin
ga2\api\zones/director-global/director/servicesets.conf: 14:1-14:36) for type 'Service' does not match anywhere!
[2019-04-03 21:03:46 +0100] warning/ApplyRule: Apply rule 'Icinga 2: cluster status' (in C:\ProgramData\icinga2\var\lib\
icinga2\api\zones/director-global/director/servicesets.conf: 25:1-25:40) for type 'Service' does not match anywhere!
[2019-04-03 21:03:46 +0100] warning/ApplyRule: Apply rule 'CPU usage' (in C:\ProgramData\icinga2\var\lib\icinga2\api\zon
es/director-global/director/servicesets.conf: 37:1-37:25) for type 'Service' does not match anywhere!
[2019-04-03 21:03:46 +0100] warning/ApplyRule: Apply rule 'Uptime' (in C:\ProgramData\icinga2\var\lib\icinga2\api\zones/
director-global/director/servicesets.conf: 45:1-45:22) for type 'Service' does not match anywhere!
[2019-04-03 21:03:46 +0100] warning/ApplyRule: Apply rule 'Memory usage' (in C:\ProgramData\icinga2\var\lib\icinga2\api\
zones/director-global/director/servicesets.conf: 53:1-53:28) for type 'Service' does not match anywhere!
[2019-04-03 21:03:46 +0100] warning/ApplyRule: Apply rule 'Uptime' (in C:\ProgramData\icinga2\var\lib\icinga2\api\zones/
director-global/director/servicesets.conf: 63:1-63:22) for type 'Service' does not match anywhere!
[2019-04-03 21:03:46 +0100] warning/ApplyRule: Apply rule 'Uptime' (in C:\ProgramData\icinga2\var\lib\icinga2\api\zones/
director-global/director/servicesets.conf: 73:1-73:22) for type 'Service' does not match anywhere!
[2019-04-03 21:03:46 +0100] warning/ApplyRule: Apply rule 'CPU usage' (in C:\ProgramData\icinga2\var\lib\icinga2\api\zon
es/director-global/director/servicesets.conf: 81:1-81:25) for type 'Service' does not match anywhere!
[2019-04-03 21:03:46 +0100] warning/ApplyRule: Apply rule 'Memory usage' (in C:\ProgramData\icinga2\var\lib\icinga2\api\
zones/director-global/director/servicesets.conf: 89:1-89:28) for type 'Service' does not match anywhere!
[2019-04-03 21:03:46 +0100] warning/ApplyRule: Apply rule 'Active sessions' (in C:\ProgramData\icinga2\var\lib\icinga2\a
pi\zones/director-global/director/servicesets.conf: 97:1-97:31) for type 'Service' does not match anywhere!
[2019-04-03 21:03:46 +0100] warning/ApplyRule: Apply rule 'Disk space: /' (in C:\ProgramData\icinga2\var\lib\icinga2\api
\zones/director-global/director/servicesets.conf: 107:1-107:29) for type 'Service' does not match anywhere!
[2019-04-03 21:03:46 +0100] warning/ApplyRule: Apply rule 'Load average' (in C:\ProgramData\icinga2\var\lib\icinga2\api\
zones/director-global/director/servicesets.conf: 118:1-118:28) for type 'Service' does not match anywhere!
[2019-04-03 21:03:46 +0100] warning/ApplyRule: Apply rule 'Memory usage' (in C:\ProgramData\icinga2\var\lib\icinga2\api\
zones/director-global/director/servicesets.conf: 129:1-129:28) for type 'Service' does not match anywhere!
[2019-04-03 21:03:46 +0100] warning/ApplyRule: Apply rule 'Load status' (in C:\ProgramData\icinga2\var\lib\icinga2\api\z
ones/director-global/director/servicesets.conf: 151:1-151:27) for type 'Service' does not match anywhere!
[2019-04-03 21:03:46 +0100] warning/ApplyRule: Apply rule 'Power usage: Watts' (in C:\ProgramData\icinga2\var\lib\icinga
2\api\zones/director-global/director/servicesets.conf: 161:1-161:34) for type 'Service' does not match anywhere!
[2019-04-03 21:03:46 +0100] warning/ApplyRule: Apply rule 'Load status: bank 1' (in C:\ProgramData\icinga2\var\lib\icing
a2\api\zones/director-global/director/servicesets.conf: 173:1-173:35) for type 'Service' does not match anywhere!
[2019-04-03 21:03:46 +0100] warning/ApplyRule: Apply rule 'Load status: bank 2' (in C:\ProgramData\icinga2\var\lib\icing
a2\api\zones/director-global/director/servicesets.conf: 182:1-182:35) for type 'Service' does not match anywhere!
[2019-04-03 21:03:46 +0100] warning/ApplyRule: Apply rule 'Identity name' (in C:\ProgramData\icinga2\var\lib\icinga2\api
\zones/director-global/director/servicesets.conf: 193:1-193:29) for type 'Service' does not match anywhere!
[2019-04-03 21:03:46 +0100] warning/ApplyRule: Apply rule 'Load' (in C:\ProgramData\icinga2\var\lib\icinga2\api\zones/di
rector-global/director/servicesets.conf: 210:1-210:20) for type 'Service' does not match anywhere!
[2019-04-03 21:03:46 +0100] warning/ApplyRule: Apply rule 'Paging File' (in C:\ProgramData\icinga2\var\lib\icinga2\api\z
ones/director-global/director/servicesets.conf: 219:1-219:27) for type 'Service' does not match anywhere!
[2019-04-03 21:03:46 +0100] warning/ApplyRule: Apply rule 'Uptime' (in C:\ProgramData\icinga2\var\lib\icinga2\api\zones/
director-global/director/servicesets.conf: 232:1-232:22) for type 'Service' does not match anywhere!
[2019-04-03 21:03:46 +0100] warning/ApplyRule: Apply rule 'Memory' (in C:\ProgramData\icinga2\var\lib\icinga2\api\zones/
director-global/director/servicesets.conf: 243:1-243:22) for type 'Service' does not match anywhere!
[2019-04-03 21:03:46 +0100] warning/ApplyRule: Apply rule 'Logged on Users' (in C:\ProgramData\icinga2\var\lib\icinga2\a
pi\zones/director-global/director/servicesets.conf: 256:1-256:31) for type 'Service' does not match anywhere!
[2019-04-03 21:03:46 +0100] warning/ApplyRule: Apply rule 'Service: VMWare Tools' (in C:\ProgramData\icinga2\var\lib\ici
nga2\api\zones/director-global/director/servicesets.conf: 265:1-265:37) for type 'Service' does not match anywhere!
[2019-04-03 21:03:46 +0100] warning/ApplyRule: Apply rule 'CPU usage' (in C:\ProgramData\icinga2\var\lib\icinga2\api\zon
es/director-global/director/servicesets.conf: 280:1-280:25) for type 'Service' does not match anywhere!
[2019-04-03 21:03:46 +0100] warning/ApplyRule: Apply rule 'Uptime' (in C:\ProgramData\icinga2\var\lib\icinga2\api\zones/
director-global/director/servicesets.conf: 288:1-288:22) for type 'Service' does not match anywhere!
[2019-04-03 21:03:46 +0100] warning/ApplyRule: Apply rule 'Memory usage' (in C:\ProgramData\icinga2\var\lib\icinga2\api\
zones/director-global/director/servicesets.conf: 296:1-296:28) for type 'Service' does not match anywhere!
[2019-04-03 21:03:46 +0100] information/ConfigItem: Instantiated 1 ApiListener.
[2019-04-03 21:03:46 +0100] information/ConfigItem: Instantiated 5 ServiceGroups.
[2019-04-03 21:03:46 +0100] information/ConfigItem: Instantiated 1 TimePeriod.
[2019-04-03 21:03:46 +0100] information/ConfigItem: Instantiated 9 HostGroups.
[2019-04-03 21:03:46 +0100] information/ConfigItem: Instantiated 1 IcingaApplication.
[2019-04-03 21:03:46 +0100] information/ConfigItem: Instantiated 3 Endpoints.
[2019-04-03 21:03:46 +0100] information/ConfigItem: Instantiated 1 FileLogger.
[2019-04-03 21:03:46 +0100] information/ConfigItem: Instantiated 89 CheckCommands.
[2019-04-03 21:03:46 +0100] information/ConfigItem: Instantiated 3 Zones.
[2019-04-03 21:03:46 +0100] information/ScriptGlobal: Dumping variables to file 'C:\ProgramData\icinga2\var\cache\icinga
2/icinga2.vars'
[2019-04-03 21:03:46 +0100] information/cli: Finished validating the configuration file(s).
```

I have attached a snippet of the log with debugging enabled.
 - 20:38:11 The Icinga2 service was started. CPU jumps to 100% after a few seconds
 - 20:38:15 I change the priority of the Icinga2 process. The CPU normalises.

[Icinga2 Log.txt](https://github.com/Icinga/icinga2/files/3040607/Icinga2.Log.txt)

**Comment by dnsmichi** [comment]
What exactly is running inside cmd.exe here? Please share the configuration of hosts/services using such custom scripts.

Besides, Windows agents consume quite a few CPU during startup when the initial local FQDN resolving doesn't work. The powershell agent script doesn't set the NodeName constant to my knowledge, the Icinga 2 setup wizard does that during the initial configuration.

Something like this inside constants.conf:

const NodeName = "windows.fqdn.domain.com"

**Comment by drewfogg** [comment]
I haven't been able to work out exactly what conhost.exe is running at the moment. The best I've been able to get is `"\??\C:\Windows\system32\conhost.exe 0x4"` from ProcessExplorer which I can't imagine is much help. I tried using [this](https://devblogs.microsoft.com/commandline/how-to-determine-what-just-ran-on-windows-console/) but it didn't work on my test machine.

I tried adding the hostname to the cosntants.conf file but it didn't seem to make a difference.

As far as I know, currently, all of our Windows services use the builtin commands and nsclient but I'll double check that tomorrow.

**Comment by drewfogg** [comment]
Hi, we don't have any custom commands/services (we're ever using the built in ones or nsclient, please could you clarify? The windows agents are all running as command endpoints. 

I've attached a copy of the debug.log (I hadn't enabled it properly last time).

The process seems to stop at the following line with high CPU:
`[2019-04-07 22:58:04 +0100] information/ApiListener: New client connection for identity '***ICINGA-Server-2***' from [::ffff:10.1.3.134]:42254`

After changing the priority of the process, the next line immediately appears (although its timestamp is the same as the line above):
`[2019-04-07 22:58:04 +0100] information/ApiListener: New client connection for identity '***ICINGA-Server-3***' from [::ffff:10.1.3.135]:50250
`

***Client01*** is a substituted name for our windows client.
***ICINGA-Server-2*** is a substituted name for our first server.
***ICINGA-Server-3*** is a substituted name for our second server.

Thanks

[icinga2_debug_log.txt](https://github.com/Icinga/icinga2/files/3052119/icinga2_debug_log.txt)

**Comment by dnsmichi** [comment]
I've talked to @LordHepipud a while ago, this was resolved within the snapshot packages for 2.11.
